dotCMS/src/main/java/com/dotcms/ai/api/OpenAIVisionAPIImpl.java

```diff
@@ -0,0 +1,398 @@
+            }
+            return Optional.of(new String(md.digest()));
+        } catch (Exception e) {
+            return Optional.empty();
+        }
+    }
+
+    @Override
+    public Optional<Tuple2<String, List<String>>> readImageTagsAndDescription(File imageFile) {
+
+        String parsedPrompt = Try.of(() -> {
+            final Context ctx = VelocityContextFactory.getMockContext();
+            ctx.put("visionModel", getAiVisionModel(Host.SYSTEM_HOST));
+            ctx.put("maxTokens", getAiVisionMaxTokens(Host.SYSTEM_HOST));
+            ctx.put("base64Image", base64EncodeImage(imageFile));
+            return VelocityUtil.eval(getAiVisionPrompt(Host.SYSTEM_HOST), ctx);
+        }).getOrNull();
+        if (parsedPrompt == null) {
+            return Optional.empty();
+        }
+
+        return readImageTagsAndDescription(parsedPrompt);
+    }
+
+
+    private Optional<Tuple2<String, List<String>>> readImageTagsAndDescription(String parsedPrompt) {
+
+        String promptHash = Try.of(
+                () -> MessageDigest.getInstance("SHA-256").digest(parsedPrompt.getBytes()).toString()).getOrNull();
+        if (UtilMethods.isEmpty(promptHash) || UtilMethods.isEmpty(parsedPrompt)) {
+            return Optional.empty();
+        }
+        return Optional.ofNullable(promptCache.get(promptHash, k -> {
+            try {
+                JSONObject parsedPromptJson = new JSONObject(parsedPrompt);
+                Logger.debug(this.getClass(), "parsedPromptJson: " + parsedPromptJson.toString());
+
+                final JSONObject openAIResponse = APILocator.getDotAIAPI()
+                        .getCompletionsAPI()
+                        .raw(parsedPromptJson, APILocator.systemUser().getUserId());
+
+                Logger.debug(OpenAIImageTaggingContentListener.class.getName(),
+                        "OpenAI Response: " + openAIResponse.toString());
+
+                final JSONObject parsedResponse = parseAIResponse(openAIResponse);
+                Logger.debug(OpenAIImageTaggingContentListener.class.getName(),
+                        "parsedResponse: " + parsedResponse.toString());
+
+                return Tuple.of(parsedResponse.getString(AI_VISION_ALT_TEXT_VARIABLE),
+                        parsedResponse.getJSONArray(AI_VISION_TAG_FIELD));
+            } catch (Exception e) {
+                Logger.warnAndDebug(OpenAIImageTaggingContentListener.class.getCanonicalName(), e.getMessage(), e);
+                return null;
+            }
+
+        }));
+
+
+    }
+
+    @Override
+    public Optional<Tuple2<String, List<String>>> readImageTagsAndDescription(Contentlet contentlet,
+            Field imageOrBinaryField) {
+
+
+
+
+        Optional<File> fileToProcess = getFileToProcess(contentlet, imageOrBinaryField);
+        if (fileToProcess.isEmpty()) {
+            return Optional.empty();
+        }
+
+        final Context ctx = VelocityContextFactory.getMockContext(contentlet, APILocator.systemUser());
```

Comment:
Not so fan of the Mock Context I would prefer that the factory realized if a request is tied in the current call and retrieve the full one, however I do not there are not too many fans of ThreadLocal approaches